realm/realm-library/src/main/cpp/io_realm_internal_objectstore_OsObjectBuilder.cpp

```diff
@@ -160,14 +160,14 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_objectstore_OsObjectBuilder_nativ
 {
     try {
         SharedRealm shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
-        Table* table = reinterpret_cast<realm::Table*>(table_ptr);
+        TableRef table = TBL_REF(table_ptr);
```

Comment:
There is a potential problem in the way a TableRef is passed to Java and back. It looses the ability to detect automatically if the table has been invalidated, because only the actual Table* from inside the TableRef is passed into Java land. When it is recreated from the Table*, it will just accept the table even if it has been recycled, and the checks in the smartptr will not fail. The correct approach is to pass the TableKey identifying the table into Java land, and then recreating the smart ptr by calling Group::get_table with the key. Alternatively, we can change the smart ptr to support being split into two JLongs and recreated from two JLongs.